synapse/logging/_structured.py

```diff
@@ -0,0 +1,268 @@
+
+    Args:
+        config (dict): A logging format v2 dictionary.
+
+    Yields:
+        HandlerConfiguration instances.
+
+    Raises:
+        ConfigError: If any of the handler configuration items are invalid.
+    """
+    for name, config in config.get("handlers").items():
+        logging_type = None
+
+        if "type" not in config and "class" in config:
+            # Handle the old "class": style.
+            if config["class"] == "logging.StreamHandler":
+                yield (
+                    HandlerConfiguration(
+                        name=name,
+                        type=LoggingOutputType.CONSOLE,
+                        location=sys.__stdout__,
+                    )
+                )
+            else:
+                raise ConfigError(
+                    "The logging class %s is not supported in logging format 2."
+                    % (config["class"],)
+                )
+        elif "type" in config:
+            logging_type = LoggingOutputType.lookupByName(config["type"].upper())
+
+            if logging_type in [LoggingOutputType.CONSOLE, LoggingOutputType.JSON]:
+                location = config.get("location")
+                if location is None or location not in ["stdout", "stderr"]:
+                    raise ConfigError(
+                        (
+                            "The %s handler needs the 'location' key set to "
+                            "either 'stdout' or 'stderr'."
+                        )
+                    )
+
+                yield HandlerConfiguration(
+                    name=name,
+                    type=logging_type,
+                    location=OutputPipeType.lookupByName(location).value,
+                )
+        else:
+            raise ConfigError("Handlers need to have either a 'type' or 'class' key.")
+
+
+def setup_structured_logging(config, log_config):
+    """
+    Set up Twisted's structured logging system.
+
+    Args:
+        config (HomeserverConfig): The configuration of the Synapse homeserver.
+        log_config (dict): The log configuration to use.
+    """
+    if config.no_redirect_stdio:
+        raise ConfigError(
+            "no_redirect_stdio cannot be defined using log_config version 2."
+        )
+
+    logger = Logger()
+
+    observers = []
+
+    for observer in parse_handler_configs(log_config):
+        if observer.type == LoggingOutputType.CONSOLE:
+            logger.debug("Starting up the {name} console logger", name=observer.name)
+            observers.append(SynapseFileLogObserver(observer.location))
+        if observer.type == LoggingOutputType.JSON:
+            logger.debug("Starting up the {name} JSON logger", name=observer.name)
+            observers.append(jsonFileLogObserver(observer.location))
+
+    publisher = LogPublisher(*observers)
+    log_filter = LogLevelFilterPredicate()
+
+    for namespace, config in log_config.get("loggers", {}).items():
+        log_filter.setLogLevelForNamespace(
```

Comment:
what is the rationale for doing this in the twisted logging layer, rather than the stdlib one?

The advantange of doing it at the stdlib layer is that if I write:

```
logger.debug("msg: %s %s %s %s", object1, object2, object3, object4)
```

then, if the logger is not configured to enable debug, the log message never actually gets formatted, and the (potentially-expensive) calls to `objectN.__str__` are not called. (I'm aware that can cause heisenbugs thanks to the different code path if you enable debugging, but on balance the ability to have `debug` statements which would be to expensive to enable on a busy server is worth it).

Reply:
I've added in filtering for the stdlib layer as well. Since the logs from Twisted structured loggers never go through the stdlib system, it needs to be set in both until we get rid of all of the stdlib loggers.

Reply:
or better, until we get rid of the twisted loggers.

Reply:
I'm not a fan of us having to do the filtering twice. Can we have separate filters specified by separate config options?

Reply:
> or better, until we get rid of the twisted loggers.

???? the entire point is to move to the Twisted logger, which is natively structured?

Reply:
this seems like a thing we need to discuss as a team.

Reply:
(This PR is about making it possible to slot in an alternative backend for the stdlib loggers, which will emit structured logs. That's a different thing to deciding to use a different *frontend* interface for our logging.)